gh-84013: normalize directory contents during import

ronaldoussoren opened the pull request:
Identifiers in Python code are normalized to NFKC and there is no guarantee that names in the file system are normalized.

Teach the FileFinder class about normalization, but only do this when the to-be-imported name contains non-ASCII characters.

<!-- gh-issue-number: gh-84013 -->
* Issue: gh-84013
<!-- /gh-issue-number -->

ronaldoussoren committed:
gh-84013: normalize directory contents during import

Identifiers in Python code are normalized to NFKC and
there is no guarantee that names in the file system are
normalized.

Teach the FileFinder class about normalization, but only
do this when the to-be-imported name contains non-ASCII
characters.

ronaldoussoren committed:
Slightly cleaner code

ronaldoussoren committed:
Argh... Previous cleanup was a bit too quick...

ronaldoussoren committed:
Actually fix the "cleaned up" version

ericsnowcurrently commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1618,12 +1619,14 @@ def find_spec(self, fullname, target=None):
         else:
-            cache = self._path_cache
             cache_module = tail_module
+            cache = self._path_cache
```
Consider leaving these lines alone.

ericsnowcurrently commented on Lib/test/test_import/__init__.py:
```diff
@@ -2701,6 +2702,43 @@ def test_pyimport_addmodule_create(self):
+        values = {}
+        exec(f"from {name} import SPAM", values, values)
+        self.assertEqual(values["SPAM"], "spam")
+        del sys.modules[unicodedata.normalize('NFKC', name)]
```
```suggestion
        filename = os.path.join(self.dir_name, f"{name}.py")
        with open(filename, "w") as stream:
            stream.write("SPAM = 'spam'\n")

        values = {}
        exec(f"from {name} import SPAM", values, values)
        try:
            self.assertEqual(values["SPAM"], "spam")
        finally:
            del sys.modules[unicodedata.normalize('NFKC', name)]
```

or even

```suggestion
        normalized = unicodedata.normalize('NFKC', name)
        filename = os.path.join(self.dir_name, f"{name}.py")
        with open(filename, "w") as stream:
            stream.write("SPAM = 'spam'\n")

        values = {}
        exec(f"from {name} import SPAM", values, values)
        try:
            self.assertEqual(values["SPAM"], "spam")
        finally:
            del sys.modules[normalized]
```

ericsnowcurrently replied:
Also, I'm not worried about `KeyError` here, since the point of this change is that we expect it to be set.  However, it might make sense to be explicit about that expectation:

```suggestion
        self.assertIn(normalized, sys.modules)
        del sys.modules[normalized]
```

(`normalized = unicodedata.normalize('NFKC', name)` is at the top of the method.)

ericsnowcurrently commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1685,6 +1688,19 @@ def _fill_cache(self):
+    def _normalize_cache(self):
+        """
+        Normalize all entries in the caches to NFKC.
+        """
```
Presumably this is not a part of `_fill_cache()` to avoid the cost of normalizing when not needed.  Is that right?

ronaldoussoren replied:
This is not part of _fill_cache for two reasons:

1. Avoid the cost of normalisation when it is not needed
2. Avoid using unicodedata during bootstrapping

ericsnowcurrently commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1618,12 +1619,14 @@ def find_spec(self, fullname, target=None):
             self._fill_cache()
             self._path_mtime = mtime
         # tail_module keeps the original casing, for __file__ and friends
+        if not tail_module.isascii() and not self._cache_is_normalized:
+            self._normalize_cache()
```
For the sake of `__import__()` and `importlib.import_module()` we will probably need to either normalize `cache_module` or preserve the original name in the cache.  The former sounds expensive so the latter might be the way to go.

ericsnowcurrently replied:
Regardless, we should probably have some test cases that use one of those functions rather than syntax.

ronaldoussoren replied:
I changed the code to keep the unnormalised name around, we need the unnormalised name to access the filesystem and need the normalised name to compare to `tail_module` (which will generally be normalised by the compiler).

I agree that there should be more test cases, I started out with just the minimal test from the issue to see if I can get this to work without bootstrapping issues and get some feedback.  I'll definitely add more tests when this approach is seen as acceptable by @brettcannon (as the importlib expert).

brettcannon replied:
If the cache always keeps the raw name then can we drop `_cache_is_normalized`?

ericsnowcurrently commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1685,6 +1688,19 @@ def _fill_cache(self):
+        from unicodedata import normalize
+
+        self._path_cache = { normalize('NFKC', p) for p in self._path_cache }
+        self._relaxed_path_cache = { normalize('NFKC', p) for p in self._relaxed_path_cache }
```
AFAICT, the cache does not influence the filename that gets returned.  As long as that filename matches what's in the cache, which is effectively a superset of allowed paths, we will get correct behavior.  That's because the code in this loader's `find_spec()` checks to make sure the actual filename (or dirname for packages) exists.

For the sake of `__import__()` and `importlib.import_module()`, we may need to preserve the un-normalized entries in the cache:

```suggestion
        self._path_cache.update(normalize('NFKC', p) for p in self._path_cache)
        self._relaxed_path_cache.update(normalize('NFKC', p) for p in self._relaxed_path_cache)
```

That's because those functions do not normalize the module name like happens for import statements.  Thus a module that currently imports via those functions could start raising `ModuleNotFoundError` because the unnormalized name isn't in the cache any more.

ericsnowcurrently replied:
There's another problem here that the PR doesn't solve: the normalized filename might not exist.  The normalized cache means the module might make it past that point.  However, the subsequent `_path_isfile(full_path)` (or `_path_isdir(base_path)`) will return `False`, leading to the same `ModuleNotFoundError` we see now.

For example:

```python
import os
name = '\u03BCmod'  # normalizes to '\u00B5'
filename = name + '.py'
with open(filename, 'w') as outfile:
    outfile.write('SPAM = 42')
ns = {}
try:
    exec(f'import {name}', ns, ns)
finally:
    os.unlink(filename)
```

Without this PR the above script fails with `ModuleNotFoundError`.  With this PR, it still fails (I'm pretty sure).  This is the same issue that PEP 3131 caused (also discussed in gh-55161).

One solution is to fall back to the known file for the relevant `_path_isfile()` and `_path_isdir()` calls.  We would store the actual path here:

```suggestion
        self._path_cache = { normalize('NFKC', p) for p in self._path_cache }
        self._relaxed_path_cache = { normalize('NFKC', p) for p in self._relaxed_path_cache }
```

(`_fill_cache()` would have to be updated to make `_path_cache` a dict with values set to `None`.)

Further, it might make sense to keep the normalized entries separate, like we do for "relaxed":

```suggestion
        self._normalized_path_cache = { normalize('NFKC', p): p for p in self._path_cache }
        self._normalized_relaxed_path_cache = { normalize('NFKC', p): p for p in self._relaxed_path_cache }
```

Then we could drop `self._cache_is_normalized` in lieu of `self._normalized_cache_path is not None`.  The fallback code in `find_spec()` would get the known actual filename from the cache and use that in place of directly using the provided module name.

That would even help somewhat with the case where there are multiple files that normalize to the same name.

ericsnowcurrently replied:
There are probably some test cases we could add to cover some of these corner cases.

ronaldoussoren replied:
> There's another problem here that the PR doesn't solve: the normalized filename might not exist. The normalized cache means the module might make it past that point. However, the subsequent _path_isfile(full_path) (or _path_isdir(base_path)) will return False, leading to the same ModuleNotFoundError we see now.

Thanks, I've been thinking about why tests pass on macOS and fail on Linux and this is likely it: Linux filesystems are not unicode-aware whereas APFS on macOS performs normalisation aware comparisons.

ericsnowcurrently replied:
Yeah, I suspected something along those lines.

ronaldoussoren replied:
Resolved by keeping both the normalized and unnormalised names around, see https://github.com/python/cpython/pull/113447#discussion_r1438171918.

ericsnowcurrently reviewed (commented):
I have some concerns.

Also, should we document that finders (both metapath-based and path-based) must deal with un-normalized names, or can we do that for all finders internally?  Instead, should we state in the language spec that all module names passing through the import system are guaranteed to be NFKC-normalized?

ronaldoussoren committed:
Change of implementation

The finder cache is now a mapping from (normalized) names
to actual filesystem names. With some luck this fixes test
failures on Linux and Windows.

ronaldoussoren committed:
Revert accidental reordering

ronaldoussoren committed:
Update Lib/test/test_import/__init__.py

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

ronaldoussoren committed:
Address review comment

ronaldoussoren committed:
Some changes

1. Add a NEWS entry

2. __import__ and find_module will now normalize their
   arguments as needed

3. Update the importlib documentation to note that the
   name passed to finders must be normalized.

ronaldoussoren commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1638,15 +1645,21 @@ def find_spec(self, fullname, target=None):
                 is_namespace = _path_isdir(base_path)
         # Check for a file w/ a proper suffix exists.
         for suffix, loader_class in self._loaders:
+            # XXX: Why is ValueError caught here?
```
This change worries me a little because I don't understand why the original code catches `ValueError` and then bails out.

Removing this like I did here does not result in test failures.

brettcannon replied:
If `git blame` isn't helpful then it can probably be removed.

serhiy-storchaka replied:
UnicodeEncodeError if the name is not compatible with the filesystem encoding, ValueError for embedded null characters?

ronaldoussoren commented on Misc/NEWS.d/next/Library/2023-12-29-13-09-44.gh-issue-84013.rUaxkr.rst:
```diff
@@ -0,0 +1,3 @@
+It is now possible to import modules from the filesystem regardless of how
+the name is normalized in the filesystem. This in particular affects
+importing modules with a name that contains accented latin characters.
```
This probably also needs a What's New entry, lets worry about that when the PR is closer to acceptance to avoid creating merge conflicts.

brettcannon commented on Lib/importlib/_bootstrap.py:
```diff
@@ -26,6 +26,13 @@ def _object_name(obj):
         return type(obj).__qualname__
 
+def _normalize(name):
+    """ Normalize 'name' in NKFC form """
```
```suggestion
    """Normalize 'name' to NKFC form."""
```

brettcannon commented on Lib/importlib/_bootstrap.py:
```diff
@@ -26,6 +26,13 @@ def _object_name(obj):
+    """ Normalize 'name' in NKFC form """
+    global _unicodedata_normalize
+    if _unicodedata_normalize is None:
+        from unicodedata import _normalize
```
```suggestion
        from unicodedata import _normalize as _unicodedata_normalize
```

brettcannon replied:
I get an import error for `_normalize` but not `normalize` in the REPL. Is that expected?

brettcannon commented on Lib/importlib/_bootstrap.py:
```diff
@@ -36,6 +43,8 @@ def _object_name(obj):
 # Import done by _install_external_importers()
 _bootstrap_external = None
 
+# Import done as needed by _normalize
```
```suggestion
# Import done lazily as needed by _normalize as unicodedata is not built-in.
```

brettcannon commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1666,24 +1679,37 @@ def _fill_cache(self):
             # the specified suffixes to check against are always specified in a
             # case-sensitive manner.
-            lower_suffix_contents = set()
+            lower_suffix_contents = dict()
```
```suggestion
            lower_suffix_contents = {}
```

brettcannon commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1666,24 +1679,37 @@ def _fill_cache(self):
+    def _normalize_cache(self):
+        """
+        Normalize all entries in the caches to NFKC.
+        """
```
```suggestion
        """Normalize all entries in the caches to NFKC."""
```

brettcannon commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -1591,8 +1591,9 @@ def __init__(self, path, *loader_details):
-        self._path_cache = set()
-        self._relaxed_path_cache = set()
+        self._path_cache = dict()
+        self._relaxed_path_cache = dict()
```
```suggestion
        self._path_cache = {}
        self._relaxed_path_cache = {}
```

serhiy-storchaka commented on Lib/importlib/_bootstrap.py:
```diff
@@ -26,6 +26,13 @@ def _object_name(obj):
+    """ Normalize 'name' in NKFC form """
+    global _unicodedata_normalize
+    if _unicodedata_normalize is None:
+        from unicodedata import _normalize
```
```suggestion
        from unicodedata import normalize as _unicodedata_normalize
```

serhiy-storchaka committed:
Apply suggestions from code review

Co-authored-by: Brett Cannon <brett@python.org>

serhiy-storchaka committed:
Merge branch 'main' into gh-84013
